### DISCUSSION TITLE
Right mechanism to inspect fields and its arguments

### DISCUSSION THREAD

**Discussion by prashant-sharma-id** [graphql-java/graphql-java#3018] [Q&A]
I am creating a service that act as a gateway service and connect multiple clients to the tenants. The tenants owns data that is accessible through the gateway service. 

The gateway service will serve data entities. Data entities will be owned by the tenants. There is a complex inter-dependency b/w data entities. I have a logic that creates the graph between the data entities and executes the graph. I am looking for the best way to inspect the asked data entities (and their arguments) from the input query. I will explain it with the example. Consider the example schema below:

type Query {
    dataEntities: DataEntities
}

type DataEntities {
   firstDataEntity: FirstDataEntity
   secondDataEntity(int testArgument): SecondDataEntity
   ... so on
}

Client can ask for multiple data entities and pass the arguments. I want to create a data fetcher for 'dataEntities' fields. In this fetcher, I want to get the name of all the entities and the passed arguments. Post that, I will run my logic: Lookup configuration of these entities, figure out their dependencies, create a graph, execute the graph. Execution of each node of the graph can be a remote call to tenant service. To that call, I need to pass arguments. 

I looked around the graphql-java documentation. I see the method 'getSelectionSet' (on DataFetchingEnvironment). It returns 'DataFetchingFieldSelectionSet' which provides simple methods to get the fields. For each field, it provides the Map of arguments also. I tested it for various scenarios (ex: argument default values, variable references etc.) and it works seamlessly. All of this was doable with very little code.

However, I wanted to get the feedback and confirm that I am going in the right direction. Please let me know if you see any concerns or a better way of doing what I am doing.

**Comment by bbakerman** [comment]
`DataFetchingFieldSelectionSet` is the best way to look ahead in a parent data fetcher.

Also look in `graphql.execution.DataFetcherResult` which allows you to send back "local context" that goes to child fields and only child fields

`graphql.execution.DataFetcherResult.Builder#localContext` will set context objects that child fields (and childs of childs) can access but its not global context and hence for all fields

**Reply by prashant-sharma-id** [reply]
Thanks for the reply. 

One follow up query:

We are using graphql-java 15.0. To get the immediate child fields, there doesn't seem to be any direct way. DataFetchingFieldSelectionSet will eventually return all the fields (child fields and their fields and so on). 

Is there a way to just grab the immediate child fields under the parent data fetcher?

One way is to use selection set and then get the field names from there. For those fields then I go to DataFetchingFieldSelectionSet and only get the requested fields (taken from the previous iteration) and then look at their arguments. However, it results in two iterations. I was wondering if there is a simplified way to do it.

**Reply by bbakerman** [reply]
`graphql.schema.DataFetchingFieldSelectionSet#getImmediateFields` gives you the immediate fields of the current selection set.

I think this method was added in version 16.

**Reply by prashant-sharma-id** [reply]
We are using version 15, so this method is not available. Thanks for the reply.
